src/state/opamSysInteract.ml

```diff
@@ -130,6 +138,15 @@ let family ~env () =
         "External dependency handling for macOS requires either \
          MacPorts or Homebrew - neither could be found"
     | "suse" | "opensuse" -> Suse
+    | "windows" ->
+      begin match OpamSysPoll.msys2_nativedir env with
+        | Some msys2_nativedir -> Msys2 (pacman_for_msys2 ~msys2_nativedir)
+        | None ->
+          failwith
+            "External dependency handling not supported for Windows unless an \
+            MSYS2 based distribution like Diskuv OCaml is installed. In \
+            particular the 'msys2-nativedir' opam var must be present"
```

Comment:
Is having this `msys2-nativedir` variable really necessary. Could we not have an automatic default detection with a fallback on `msys2-nativedir`?
Looking at the msys2 documentation, for example for users that is already using the Mintty terminal, `pacman` should already be available. Couldn't we have a check where `if OpamSystem.resolve_command "pacman"` then we just use msys2 by default, like homebrew vs. macports on macOS?

Reply:
> Couldn't we have a check where `if OpamSystem.resolve_command "pacman"` then we just use msys2 by default, like homebrew vs. macports on macOS?

This is the central question that I think affects the other questions you have in the PR.

Short story: We could check for `pacman` in the PATH, but that should not be the primary way that MSYS2 is detected. Why? Because **A)** Windows users would have poor user experiences if they were forced to open a Mintty terminal to run opam (or ocaml or dune, etc.), and **B)** there can be multiple `pacman` on a machine and because **C)** MSYS2 is a temporary hack.

Long long story below ... which probably should go in a README or a ROADMAP somewhere ...

#### A - Poor Windows experience

In my Testing section you may have noticed I had to do `with-dkml` before I ran opam. `with-dkml` replaces opening up a Mintty terminal, and does a bunch of other things not relevant to this PR. Since having to type `with-dkml` in front of every `opam` (or `dune`) command would a) frustrate Windows users, b) require package maintainers to complicate their README.md and odoc instructions, and c) cause me a huge support headache, I have provided shims as a temporary solution.

Shim: The `opam.exe` in the PATH is really a special version of `with-dkml.exe` that does all the environment manipulation that Mintty would do, but then searches for and spawns the real opam binary.

The Windows user can now run `opam` and `dune` in the two shells they are already familiar with: the Command Prompt and PowerShell. That gives Windows users the same experience as a Linux or macOS user (who are not told to run a special shell just to use `opam`).

The real end-goal I have been walking towards is that plain old `opam.exe` compiled from the master branch of github.com/ocaml/opam works on Windows in the Command Prompt and PowerShell with no shim required.

#### B - Multiple MSYS2 on machine

It is common for MSYS2 to be embedded multiple times on one machine. For example, most Windows developers have Git for Windows which bundles its own MSYS2 and also its own terminal "Git Bash". And many Windows C developers use Microsoft's C package manager (vcpkg) which also bundles its own MSYS2.  So it is important that OCaml depexts packages are added to the correct MSYS2 installation. That is why you see a full path to `pacman.exe` in this PR; I don't, for example, want to pick up the vcpkg `pacman.exe` accidentally.

#### C - MSYS2 is a temporary hack

There are three reasons why MSYS2 is required:

1. Today `opam.exe` compiled from the master branch spawns Unix tools like `cp`. I heard that your team will be fixing that at some point (probably replacing those Unix tools with pure OCaml code). The opam shim would be unnecessary once that is done! PS. My [diskuvbox library](https://github.com/diskuv/diskuvbox#readme) may be of help here. 
2. Some opam packages use Unix tools in their `build:[]` instructions. I use `with-dkml.exe` again as a hack by setting the opam `wrap-build-commands` (etc.) options in their switches. This is bad though ... Windows users cannot simply create a switch with `opam switch create` but now they need `opam dkml init` to ensure the correct opam options are set.
3. Compiling `ocaml.exe` and `ocamlc.exe` (etc.) requires autoconf, which requires a POSIX environment like MSYS2 or Cygwin. This is solvable (aka. Bazel and CMake) although this is low priority. More importantly, if this was the only need for MSYS2 it could be isolated to just `ocaml-base-compiler`.

The opam shim and the wrap-build-commands options are hacks and I'd like them to disappear as fast as humanly possible.